src/main/java/com/minecolonies/coremod/commands/AbstractCommandParser.java

```diff
@@ -0,0 +1,649 @@
+        {
+            throw new CommandException(getCommandUsage(sender, getRoot()));
+        }
+
+        final IMenu executionMenu = executionTreeNode.getData();
+        if (executionMenu.getMenuType().isNavigationMenu())
+        {
+            throw new CommandException(getCommandUsage(sender, executionTreeNode));
+        }
+
+        @NotNull final List<ActionArgument> executionActionArgumentList = parsingResult.getExecutionActionArgumentList();
+        // actionMenuState will be non-null at this point
+        @NotNull final ActionMenuState actionMenuState = parsingResult.getActionMenuState();
+        final ActionMenu actionMenu = actionMenuState.getActionMenu();
+
+        final String badArgument = parsingResult.getBadArgument();
+        throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(executionTreeNode, actionMenu, executionActionArgumentList, actionMenuState, badArgument, sender);
+
+        if (sender instanceof EntityPlayer)
+        {
+            final ForgePermissionNodes forgePermissionNode = actionMenu.getForgePermissionNode();
+
+            final EntityPlayer player = (EntityPlayer) sender;
+            if (!permissionsChecker.hasPermission(forgePermissionNode, player))
+            {
+                // TODO: Do something if permission check fails.
+                // But we don't have permissions set up yet.
+            }
+        }
+
+        for (final ActionArgument executionActionArgument : executionActionArgumentList)
+        {
+            if (!actionMenuState.isValueSet(executionActionArgument))
+            {
+                throw new CommandException(getCommandUsage(sender, executionTreeNode));
+            }
+        }
+
+        final Class<? extends IActionCommand> clazz = actionMenu.getActionCommandClass();
+        try
+        {
+            createInstanceAndExecute(server, sender, actionMenuState, clazz);
+        }
+        catch (final InstantiationException | IllegalAccessException e)
+        {
+            final Logger log = LogManager.getLogger();
+            log.error("Unable to instantiate class %s for command %s ", clazz.getName(), actionMenu.getDescription(), e);
+            throw new CommandException("Unable to instantiate class " + clazz.getName() + " for command " + actionMenu.getDescription(), e);
+        }
+    }
+
+    protected void createInstanceAndExecute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenuState actionMenuState,
+            @NotNull final Class<? extends IActionCommand> clazz) throws InstantiationException, IllegalAccessException, CommandException
+    {
+        final IActionCommand actionCommand = clazz.newInstance();
+        actionCommand.execute(server, sender, actionMenuState);
+    }
+
+    private static void throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(@NotNull final TreeNode<IMenu> executionTreeNode, @NotNull final ActionMenu actionMenu,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 34 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)